crates/ironrdp-rdpsnd-native/src/cpal.rs

```diff
@@ -0,0 +1,244 @@
+use std::{
+    borrow::Cow,
+    sync::mpsc::{self, Receiver, Sender},
+    thread::{self, JoinHandle},
+    time::Duration,
+};
+
+use anyhow::{bail, Context};
+use cpal::{
+    traits::{DeviceTrait, HostTrait},
+    SampleFormat, Stream, StreamConfig,
+};
+use ironrdp_rdpsnd::{
+    client::RdpsndClientHandler,
+    pdu::{AudioFormat, PitchPdu, VolumePdu, WaveFormat},
+};
+use tracing::{debug, error, info};
+
+#[derive(Debug)]
+pub struct RdpsndBackend {
+    stream_handle: Option<JoinHandle<()>>,
+    tx: Option<Sender<Vec<u8>>>,
+    format: Option<AudioFormat>,
+}
+
+impl Default for RdpsndBackend {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl RdpsndBackend {
+    pub fn new() -> Self {
+        Self {
+            tx: None,
+            format: None,
+            stream_handle: None,
+        }
+    }
+}
+
+impl RdpsndClientHandler for RdpsndBackend {
+    fn wave(&mut self, format: &AudioFormat, _ts: u32, data: Cow<'_, [u8]>) {
+        if Some(format) != self.format.as_ref() {
+            debug!("new audio format {format:?}");
+            if let Some(ref stream) = self.stream_handle {
+                stream.thread().unpark();
+                self.stream_handle = None;
+            }
+        }
+
+        if self.stream_handle.is_none() {
+            let (tx, rx) = mpsc::channel();
+            self.tx = Some(tx);
+            self.format = Some(format.clone());
+            let format = format.clone();
+            self.stream_handle = Some(thread::spawn(move || {
+                let _stream = match make_stream(&format, rx) {
+                    Ok(stream) => Some(stream),
+                    Err(err) => {
+                        error!("{}", err);
+                        return;
+                    }
+                };
+                debug!("parking");
+                thread::park();
+                debug!("unparking");
```

Comment:
question: Can you elaborate with a comment on why we need to park the thread manually once before it terminates, please?

Reply:
done, and improved the code